planetiler-core/src/main/java/com/onthegomap/planetiler/pmtiles/WriteablePmtiles.java

```diff
@@ -0,0 +1,619 @@
+    }
+
+    for (int i = 0; i < numEntries; i++) {
+      long tmp = VarInt.getVarLong(buffer);
+      if (i > 0 && tmp == 0) {
+        result.get(i).offset = result.get(i - 1).offset + result.get(i - 1).length;
+      } else {
+        result.get(i).offset = tmp - 1;
+      }
+    }
+    return result;
+  }
+
+  /**
+   * Arbitrary application-specific JSON metadata in the archive.
+   * <p>
+   * stores name, attribution, created_at, planetiler build SHA, vector_layers, etc.
+   */
+  public record JsonMetadata(
+    @JsonProperty("vector_layers") List<LayerStats.VectorLayer> vectorLayers,
+    @JsonAnyGetter Map<String, String> otherMetadata
+  ) {
+
+    public byte[] toBytes() {
+      try {
+        return objectMapper.writeValueAsBytes(this);
+      } catch (JsonProcessingException e) {
+        throw new IllegalArgumentException("Unable to encode as string: " + this, e);
+      }
+    }
+  }
+
+  public record Directories(byte[] root, byte[] leaves, int numLeaves) {}
+
+  private static Directories buildRootLeaves(List<Entry> subEntries, int leafSize) throws IOException {
+    ArrayList<Entry> rootEntries = new ArrayList<>();
+    ByteArrayList leavesOutputStream = new ByteArrayList();
+    int leavesLength = 0;
+    int numLeaves = 0;
+
+    for (int i = 0; i < subEntries.size(); i += leafSize) {
+      numLeaves++;
+      int end = i + leafSize;
+      if (i + leafSize > subEntries.size()) {
+        end = subEntries.size();
+      }
+      byte[] leafBytes = serializeDirectory(subEntries, i, end);
+      leafBytes = Gzip.gzip(leafBytes);
+      rootEntries.add(new Entry(subEntries.get(i).tileId, leavesLength, leafBytes.length, 0));
+      leavesOutputStream.add(leafBytes);
+      leavesLength += leafBytes.length;
+    }
+
+    byte[] rootBytes = serializeDirectory(rootEntries, 0, rootEntries.size());
+    rootBytes = Gzip.gzip(rootBytes);
+
+    return new Directories(rootBytes, leavesOutputStream.toArray(), numLeaves);
+  }
+
+  /**
+   * Serialize all entries into bytes, choosing the # of leaf directories to ensure the header+root fits in 16 KB.
+   *
+   * @param entries a sorted ObjectArrayList of all entries in the tileset.
+   * @return byte arrays of the root and all leaf directories, and the # of leaves.
+   * @throws IOException
+   */
+  public static Directories makeRootLeaves(List<Entry> entries) throws IOException {
+    if (entries.size() < 16384) { // coarse heuristic to short circuit
+      byte[] testBytes = serializeDirectory(entries, 0, entries.size());
+      testBytes = Gzip.gzip(testBytes);
+
+      if (testBytes.length < INIT_SECTION - HEADER_LEN) {
+        return new Directories(testBytes, new byte[0], 0);
+      }
+    }
+
+    int leafSize = (int) Math.max(entries.size() / 3_500d, 4096);
+
+    while (true) {
+      Directories temp = buildRootLeaves(entries, leafSize);
+      if (temp.root.length < INIT_SECTION - HEADER_LEN) {
+        return temp;
+      }
+      leafSize *= 1.2;
+    }
+  }
```

Comment:
I don't think so, the `leafSize` will grow towards infinity, possibly slowly. If we used a numerical-ish method to attempt to converge on exactly 16KB header, there would be some risk of that. The current strategy is to use some rough numbers based on typical real data to get around 16KB on first attempt, and then refine via `1.2` till we get under. 

The overall purpose here is to avoid situations where we have undersized roots like 8kB; we could ~double the # of leaves and make each individual leaf smaller and faster to fetch, while not affecting the overall latency since we blindly fetch 16KB in most clients anyway.

Reply:
A bit difficult to unit test the adaptive sizing because those specific parameters are cherry-picked out of real-world data = how compressible the directories are; we can refine those parameters later based on real datasets